<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3: Generalized Linear Models | Analysis Using R 2024</title>
  <meta name="description" content="Workshop website for the 2024 Analysis Using R Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3: Generalized Linear Models | Analysis Using R 2024" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://CBWGitHub/AUR-2024/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the 2024 Analysis Using R Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="cbw-dev/AUR_2024" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3: Generalized Linear Models | Analysis Using R 2024" />
  
  <meta name="twitter:description" content="Workshop website for the 2024 Analysis Using R Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://CBWGitHub/AUR-2024/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Shraddha Pai, Chaitra Sarathy, Ian Cheong, Zoe Klein, Amin Nooranikhojasteh, Nia Hughes" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.icon" type="image/x-icon" />
<link rel="prev" href="module-2-dimensionality-reduction.html"/>
<link rel="next" href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
 <li><center><strong><a href="./"><br>Analysis Using R<br><br></a></strong></center></li>
 <a href="./">
 <img src="img/AUR_icon.png" width="70%" alt="Analysis Using R logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
 <br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#lecture"><i class="fa fa-check"></i>Lecture</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#downloads"><i class="fa fa-check"></i>Downloads</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#lab"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#bonus-exercise"><i class="fa fa-check"></i>Bonus Exercise</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#bonus-exercise-results"><i class="fa fa-check"></i>Bonus Exercise Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality Reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#lecture-1"><i class="fa fa-check"></i>Lecture</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#downloads-1"><i class="fa fa-check"></i>Downloads</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#bonus-exercise-1"><i class="fa fa-check"></i>Bonus Exercise</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#bonus-exercise-results-1"><i class="fa fa-check"></i>Bonus Exercise Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#lecture-2"><i class="fa fa-check"></i>Lecture</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#downloads-2"><i class="fa fa-check"></i>Downloads</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#bonus-exercise-2"><i class="fa fa-check"></i>Bonus Exercise</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#module-3-bonus-exercise-results"><i class="fa fa-check"></i>Module 3: Bonus Exercise Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding Differentially Expressed Genes From RNAseq Data</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#lecture-3"><i class="fa fa-check"></i>Lecture</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#downloads-3"><i class="fa fa-check"></i>Downloads</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#volcano-plot-r-base-graphics"><i class="fa fa-check"></i>Volcano plot (R base graphics)</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#volcano-plot-ggplot2"><i class="fa fa-check"></i>Volcano plot (ggplot2)</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#bonus-exercise-3"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
<br><br>
<a href="https://www.bioinformatics.ca"><img src="img/sponsors/c3g_logo.png" style="display:inline-block;" width=35% style="padding-bottom: 20px" alt="Canadian Centre for Computational Genomics logo."> <!--- replace the directory with your sponsor logo --->


<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li></center>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R 2024</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3-generalized-linear-models" class="section level1 hasAnchor">
<h1>Module 3: Generalized Linear Models<a href="module-3-generalized-linear-models.html#module-3-generalized-linear-models" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-2" class="section level2 hasAnchor">
<h2>Lecture<a href="module-3-generalized-linear-models.html#lecture-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><embed src="content-files/AUR_2024_Module3_GLM.pdf" style="width:100.0%" height="900" /> <!-- edit height based your preferences --></p>
<div id="downloads-2" class="section level3 hasAnchor">
<h3>Downloads<a href="module-3-generalized-linear-models.html#downloads-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>[insert your downloads for this module here (ex. datasets)]</p>
</div>
</div>
<div id="lab-2" class="section level2 hasAnchor">
<h2>Lab<a href="module-3-generalized-linear-models.html#lab-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this module we’re going to cover:</p>
<ul>
<li>Reading data from files and evaluating missingness</li>
<li>Creating publication-quality plots with <code>ggplot2</code></li>
<li>Fitting models with binary outcomes using generalized linear models</li>
</ul>
<div id="essential-r-reading-tables-from-files-merging-basic-data-exploration" class="section level3 unnumbered hasAnchor">
<h3>Essential R: Reading tables from files, merging, basic data exploration<a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For this we’re going to use two data files available in the course data directory.</p>
<p><a href="https://drive.google.com/drive/folders/1hKRa7ui-DmQCbqoKlgkmvhlrbXDmbgsV?usp=sharing">Download these</a> for use and put them in your R working directory.</p>
<ul>
<li><code>tomerge1.csv</code> : comma-separated values</li>
<li><code>tomerge2.txt</code> : space-delimited values</li>
</ul>
<p>Use <code>read.delim()</code> to read in tables. Note the use of the <code>sep=</code> parameter to indicate what the column separator is:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb222-1"><a href="module-3-generalized-linear-models.html#cb222-1" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;AUR2024_data/tomerge1.csv&quot;</span>,<span class="at">sep=</span><span class="st">&quot;,&quot;</span>)</span>
<span id="cb222-2"><a href="module-3-generalized-linear-models.html#cb222-2" tabindex="-1"></a><span class="fu">head</span>(x1)</span></code></pre></div>
<pre><code>##   Sample_ID Exposure Biomarker_value
## 1         A        0              35
## 2         B        0              22
## 3         C        0              91
## 4         D        0               3
## 5         E        1              56
## 6         F        1              37</code></pre>
<div class="sourceCode" id="cb224"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb224-1"><a href="module-3-generalized-linear-models.html#cb224-1" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;AUR2024_data/tomerge2.txt&quot;</span>,<span class="at">sep=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb224-2"><a href="module-3-generalized-linear-models.html#cb224-2" tabindex="-1"></a><span class="fu">head</span>(x2)</span></code></pre></div>
<pre><code>##   sampleID Exposure_level Biomarker2_detected
## 1        A     0.65405517                   1
## 2        B     0.67202852                   0
## 3        C     0.88646372                   1
## 4        D     0.28433256                   0
## 5        E     0.04166839                   0
## 6        F     0.45263534                   1</code></pre>
<p>Use <code>merge()</code> to combine the two tables by sample ID. Note the use of <code>by.x</code> and <code>by.y</code> to tell <code>merge()</code> which columns are equivalent:</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb226-1"><a href="module-3-generalized-linear-models.html#cb226-1" tabindex="-1"></a>x_merge <span class="ot">&lt;-</span> <span class="fu">merge</span>(x1, x2, <span class="at">by.x =</span> <span class="st">&quot;Sample_ID&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;sampleID&quot;</span>)</span>
<span id="cb226-2"><a href="module-3-generalized-linear-models.html#cb226-2" tabindex="-1"></a><span class="fu">head</span>(x_merge)</span></code></pre></div>
<pre><code>##   Sample_ID Exposure Biomarker_value Exposure_level Biomarker2_detected
## 1         A        0              35     0.65405517                   1
## 2         B        0              22     0.67202852                   0
## 3         C        0              91     0.88646372                   1
## 4         D        0               3     0.28433256                   0
## 5         E        1              56     0.04166839                   0
## 6         F        1              37     0.45263534                   1</code></pre>
<p>We’re going to use a popular dataset for data analysis, pertaining to survival of passengers aboard the Titanic.</p>
<p>Download <a href="https://drive.google.com/file/d/1PdP7bZTbmcRrPPPYFf1Mr8vfucom-d4k/view?usp=sharing">the dataset here</a> and copy it into your working directory.</p>
<p>Let’s read in the data from a file:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb228-1"><a href="module-3-generalized-linear-models.html#cb228-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(</span>
<span id="cb228-2"><a href="module-3-generalized-linear-models.html#cb228-2" tabindex="-1"></a>    <span class="st">&quot;AUR2024_data/titanic.csv&quot;</span>,</span>
<span id="cb228-3"><a href="module-3-generalized-linear-models.html#cb228-3" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">&quot;,&quot;</span> <span class="co"># indicate the column separator</span></span>
<span id="cb228-4"><a href="module-3-generalized-linear-models.html#cb228-4" tabindex="-1"></a>    ) </span></code></pre></div>
<p>Examine the columns:</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb229-1"><a href="module-3-generalized-linear-models.html#cb229-1" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   PassengerId Survived Pclass
## 1           1        0      3
## 2           2        1      1
## 3           3        1      3
## 4           4        1      1
## 5           5        0      3
## 6           6        0      3
##                                                  Name    Sex Age SibSp Parch
## 1                             Braund, Mr. Owen Harris   male  22     1     0
## 2 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female  38     1     0
## 3                              Heikkinen, Miss. Laina female  26     0     0
## 4        Futrelle, Mrs. Jacques Heath (Lily May Peel) female  35     1     0
## 5                            Allen, Mr. William Henry   male  35     0     0
## 6                                    Moran, Mr. James   male  NA     0     0
##             Ticket    Fare Cabin Embarked
## 1        A/5 21171  7.2500              S
## 2         PC 17599 71.2833   C85        C
## 3 STON/O2. 3101282  7.9250              S
## 4           113803 53.1000  C123        S
## 5           373450  8.0500              S
## 6           330877  8.4583              Q</code></pre>
<p>Some of the columns are categorical, use <code>table()</code> to look at the tallies:
Examine the columns:</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb231-1"><a href="module-3-generalized-linear-models.html#cb231-1" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Survived)</span></code></pre></div>
<pre><code>## 
##   0   1 
## 549 342</code></pre>
<div class="sourceCode" id="cb233"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb233-1"><a href="module-3-generalized-linear-models.html#cb233-1" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Pclass)</span></code></pre></div>
<pre><code>## 
##   1   2   3 
## 216 184 491</code></pre>
<p>Use <code>summary()</code> to look at continuous-valued data:</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb235-1"><a href="module-3-generalized-linear-models.html#cb235-1" tabindex="-1"></a><span class="fu">summary</span>(dat<span class="sc">$</span>Age)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
##    0.42   20.12   28.00   29.70   38.00   80.00     177</code></pre>
<p>Notice that there are 177 <code>NA</code> (missing) values for age.
Let’s visualize the missing data more systematically.</p>
</div>
<div id="explore-missing-data" class="section level3 unnumbered hasAnchor">
<h3>Explore missing data<a href="module-3-generalized-linear-models.html#explore-missing-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For this let’s use a little script that converts a table into black and white squares to visualize missing data. For this, install the <code>plotrix</code> package</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb237-1"><a href="module-3-generalized-linear-models.html#cb237-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;plotrix&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;plotrix&quot;</span>)</span>
<span id="cb237-2"><a href="module-3-generalized-linear-models.html#cb237-2" tabindex="-1"></a></span>
<span id="cb237-3"><a href="module-3-generalized-linear-models.html#cb237-3" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(plotrix))</span>
<span id="cb237-4"><a href="module-3-generalized-linear-models.html#cb237-4" tabindex="-1"></a></span>
<span id="cb237-5"><a href="module-3-generalized-linear-models.html#cb237-5" tabindex="-1"></a><span class="co">#&#39; show data missingness as a chequered matrix</span></span>
<span id="cb237-6"><a href="module-3-generalized-linear-models.html#cb237-6" tabindex="-1"></a><span class="co">#&#39; </span></span>
<span id="cb237-7"><a href="module-3-generalized-linear-models.html#cb237-7" tabindex="-1"></a><span class="co">#&#39; @param x (matrix) data matrix.</span></span>
<span id="cb237-8"><a href="module-3-generalized-linear-models.html#cb237-8" tabindex="-1"></a><span class="co">#&#39; @param outFile (char) path to file for printing graph</span></span>
<span id="cb237-9"><a href="module-3-generalized-linear-models.html#cb237-9" tabindex="-1"></a><span class="co">#&#39; @param wd (numeric) width in inches</span></span>
<span id="cb237-10"><a href="module-3-generalized-linear-models.html#cb237-10" tabindex="-1"></a><span class="co">#&#39; @param ht (numeric) height in inches</span></span>
<span id="cb237-11"><a href="module-3-generalized-linear-models.html#cb237-11" tabindex="-1"></a><span class="co">#&#39; @return plots missingness matrix to file</span></span>
<span id="cb237-12"><a href="module-3-generalized-linear-models.html#cb237-12" tabindex="-1"></a><span class="co">#&#39; @import plotrix</span></span>
<span id="cb237-13"><a href="module-3-generalized-linear-models.html#cb237-13" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb237-14"><a href="module-3-generalized-linear-models.html#cb237-14" tabindex="-1"></a>plotMissMat <span class="ot">&lt;-</span> <span class="cf">function</span>(x,<span class="at">xlab=</span><span class="st">&quot;columns&quot;</span>,</span>
<span id="cb237-15"><a href="module-3-generalized-linear-models.html#cb237-15" tabindex="-1"></a>        <span class="at">ylab=</span><span class="st">&quot;rows&quot;</span>,<span class="at">border=</span><span class="cn">NA</span>) {</span>
<span id="cb237-16"><a href="module-3-generalized-linear-models.html#cb237-16" tabindex="-1"></a>    </span>
<span id="cb237-17"><a href="module-3-generalized-linear-models.html#cb237-17" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(x)</span>
<span id="cb237-18"><a href="module-3-generalized-linear-models.html#cb237-18" tabindex="-1"></a>    <span class="fu">class</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;numeric&quot;</span></span>
<span id="cb237-19"><a href="module-3-generalized-linear-models.html#cb237-19" tabindex="-1"></a>    <span class="fu">color2D.matplot</span>(x,<span class="at">show.values=</span><span class="cn">FALSE</span>,<span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb237-20"><a href="module-3-generalized-linear-models.html#cb237-20" tabindex="-1"></a>        <span class="at">cs1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs2=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs3=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">border=</span>border,</span>
<span id="cb237-21"><a href="module-3-generalized-linear-models.html#cb237-21" tabindex="-1"></a>        <span class="at">cex=</span><span class="fl">0.8</span>,</span>
<span id="cb237-22"><a href="module-3-generalized-linear-models.html#cb237-22" tabindex="-1"></a>        <span class="at">xlab=</span>xlab,<span class="at">ylab=</span>ylab)</span>
<span id="cb237-23"><a href="module-3-generalized-linear-models.html#cb237-23" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s look at the missingness in the Titanic dataset. Missing data is shown as a white cell, and non-missing data is shown in black.</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb238-1"><a href="module-3-generalized-linear-models.html#cb238-1" tabindex="-1"></a><span class="fu">plotMissMat</span>(dat)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-70-1.png" width="672" /></p>
<p>We can see a column with many missing values. This is probably the “age” data.
Let’s count the number of missing values on a per-column level.</p>
<p>For this we combine <code>is.na()</code>, which returns a TRUE/FALSE value for NA values, and <code>colSums()</code> which adds up the TRUE values down the columns.</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb239-1"><a href="module-3-generalized-linear-models.html#cb239-1" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(dat))</span></code></pre></div>
<pre><code>## PassengerId    Survived      Pclass        Name         Sex         Age 
##           0           0           0           0           0         177 
##       SibSp       Parch      Ticket        Fare       Cabin    Embarked 
##           0           0           0           0           0           0</code></pre>
<p>This confirms that <code>Age</code> is the only column with missing data.</p>
<p>Now let’s explore the data using plots.</p>
</div>
<div id="essential-r-plots-with-ggplot2" class="section level3 unnumbered hasAnchor">
<h3>Essential R: Plots with <code>ggplot2</code><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>ggplot2</code> is a popular plotting package that uses an additive paradigm to build plots.</p>
<p>Useful websites:
* <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">ggplot2 cheatsheet</a>
* The <a href="https://ggplot2.tidyverse.org/">ggplot2 website</a> is a wealth of reference information, so here we just touch on the basics to get you started.</p>
<p>Anytime you need to generate a specific kind of plot, the website will most likely have documentation for how to achieve it.</p>
<p>Let’s start by creating a scatterplot with two <strong>continuous variables</strong>. For this let’s load a dataset measuring statistics around quality of life in US states in the late 70’s:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb241-1"><a href="module-3-generalized-linear-models.html#cb241-1" tabindex="-1"></a>state.x77 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(state.x77)</span>
<span id="cb241-2"><a href="module-3-generalized-linear-models.html#cb241-2" tabindex="-1"></a><span class="fu">head</span>(state.x77)</span></code></pre></div>
<pre><code>##            Population Income Illiteracy Life Exp Murder HS Grad Frost   Area
## Alabama          3615   3624        2.1    69.05   15.1    41.3    20  50708
## Alaska            365   6315        1.5    69.31   11.3    66.7   152 566432
## Arizona          2212   4530        1.8    70.55    7.8    58.1    15 113417
## Arkansas         2110   3378        1.9    70.66   10.1    39.9    65  51945
## California      21198   5114        1.1    71.71   10.3    62.6    20 156361
## Colorado         2541   4884        0.7    72.06    6.8    63.9   166 103766</code></pre>
<p>Create a base plot using the <code>ggplot</code> function. Then “add” a scatterplot to it. Notice that the plot has been assigned to a variable named <code>p</code>.</p>
<p>This setup is standard for <code>ggplot2</code> and allows multiple visualizations to be applied to the same base plot.</p>
<p>We use <code>aes</code> to tell ggplot what the <code>x</code> and <code>y</code> axes are, and later, if we want to colour-code by a particular column.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb243-1"><a href="module-3-generalized-linear-models.html#cb243-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb243-2"><a href="module-3-generalized-linear-models.html#cb243-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(state.x77, </span>
<span id="cb243-3"><a href="module-3-generalized-linear-models.html#cb243-3" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> Illiteracy,<span class="at">y =</span> Income)</span>
<span id="cb243-4"><a href="module-3-generalized-linear-models.html#cb243-4" tabindex="-1"></a>    )</span>
<span id="cb243-5"><a href="module-3-generalized-linear-models.html#cb243-5" tabindex="-1"></a></span>
<span id="cb243-6"><a href="module-3-generalized-linear-models.html#cb243-6" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="co"># scatter plot</span></span>
<span id="cb243-7"><a href="module-3-generalized-linear-models.html#cb243-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="_main_files/figure-html/load-states-data-1.png" width="672" /></p>
<p>Now let’s add confidence intervals:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb244-1"><a href="module-3-generalized-linear-models.html#cb244-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<p><img src="_main_files/figure-html/states-geomsmooth-1.png" width="672" /></p>
<p>It looks like there is a negative relationship between illiteracy and income.
We can confirm this by looking at correlation:</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb246-1"><a href="module-3-generalized-linear-models.html#cb246-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Illiteracy</span>
<span id="cb246-2"><a href="module-3-generalized-linear-models.html#cb246-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Income</span>
<span id="cb246-3"><a href="module-3-generalized-linear-models.html#cb246-3" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  x and y
## t = -3.3668, df = 48, p-value = 0.001505
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.6378257 -0.1807128
## sample estimates:
##        cor 
## -0.4370752</code></pre>
<div class="sourceCode" id="cb248"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb248-1"><a href="module-3-generalized-linear-models.html#cb248-1" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)<span class="sc">$</span>p.value</span></code></pre></div>
<pre><code>## [1] 0.001505073</code></pre>
<p>Let’s now examine <strong>categorical variables</strong>.
In the titanic set, let’s look at the fare paid based on passenger class:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb250-1"><a href="module-3-generalized-linear-models.html#cb250-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat)</span>
<span id="cb250-2"><a href="module-3-generalized-linear-models.html#cb250-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_boxplot</span>(</span>
<span id="cb250-3"><a href="module-3-generalized-linear-models.html#cb250-3" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(Pclass), <span class="co"># &quot;factor()&quot; makes a data column a categorical variable</span></span>
<span id="cb250-4"><a href="module-3-generalized-linear-models.html#cb250-4" tabindex="-1"></a>        <span class="at">y =</span> Fare))</span></code></pre></div>
<p><img src="_main_files/figure-html/boxplot-1.png" width="672" /></p>
<p>We can use barplots to examine counts and proportions.
Let’s look at number of survivors, split by passenger class.</p>
<p>Here the <code>fill</code> command is used to to split each barplot by the category, “Survived”.
So you can see number of passengers by “Pclass” split by survival.</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb251-1"><a href="module-3-generalized-linear-models.html#cb251-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb251-2"><a href="module-3-generalized-linear-models.html#cb251-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), </span>
<span id="cb251-3"><a href="module-3-generalized-linear-models.html#cb251-3" tabindex="-1"></a>        <span class="at">x =</span> Pclass)</span>
<span id="cb251-4"><a href="module-3-generalized-linear-models.html#cb251-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="_main_files/figure-html/barplot-1.png" width="672" /></p>
<p>The plot above shows count data. Let’s convert this to proportions. We can see that the fraction of non-survivors in “Class 3” is high.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb252-1"><a href="module-3-generalized-linear-models.html#cb252-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb252-2"><a href="module-3-generalized-linear-models.html#cb252-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), <span class="at">x =</span> Pclass),</span>
<span id="cb252-3"><a href="module-3-generalized-linear-models.html#cb252-3" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb252-4"><a href="module-3-generalized-linear-models.html#cb252-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="_main_files/figure-html/barplot2%20-1.png" width="672" /></p>
<p>How about males versus females?</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb253-1"><a href="module-3-generalized-linear-models.html#cb253-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb253-2"><a href="module-3-generalized-linear-models.html#cb253-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), <span class="at">x =</span> Sex),</span>
<span id="cb253-3"><a href="module-3-generalized-linear-models.html#cb253-3" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb253-4"><a href="module-3-generalized-linear-models.html#cb253-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="_main_files/figure-html/finish-titanic-1.png" width="672" /></p>
<p><strong>Exercise for later:</strong> Try other <code>ggplot</code> functions on these data or on a small data table of your project.
(avoid using large genomics datasets, because those are going to be harder to interpret)</p>
</div>
<div id="fit-binary-response-variable-using-glm-and-logistic-regression" class="section level3 unnumbered hasAnchor">
<h3>Fit binary response variable using <code>glm()</code> and logistic regression<a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s fit a model to a binary outcome. For this we load a dataset that measures physiological variables in a cohort of Pima Indians.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb254-1"><a href="module-3-generalized-linear-models.html#cb254-1" tabindex="-1"></a><span class="fu">library</span>(mlbench)</span>
<span id="cb254-2"><a href="module-3-generalized-linear-models.html#cb254-2" tabindex="-1"></a><span class="fu">data</span>(PimaIndiansDiabetes2)</span>
<span id="cb254-3"><a href="module-3-generalized-linear-models.html#cb254-3" tabindex="-1"></a><span class="co"># type ?PimaIndiansDiabetes2 to learn more about the dataset.</span></span>
<span id="cb254-4"><a href="module-3-generalized-linear-models.html#cb254-4" tabindex="-1"></a></span>
<span id="cb254-5"><a href="module-3-generalized-linear-models.html#cb254-5" tabindex="-1"></a>dat <span class="ot">&lt;-</span> PimaIndiansDiabetes2</span>
<span id="cb254-6"><a href="module-3-generalized-linear-models.html#cb254-6" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   pregnant glucose pressure triceps insulin mass pedigree age diabetes
## 1        6     148       72      35      NA 33.6    0.627  50      pos
## 2        1      85       66      29      NA 26.6    0.351  31      neg
## 3        8     183       64      NA      NA 23.3    0.672  32      pos
## 4        1      89       66      23      94 28.1    0.167  21      neg
## 5        0     137       40      35     168 43.1    2.288  33      pos
## 6        5     116       74      NA      NA 25.6    0.201  30      neg</code></pre>
<p>Let’s look at the impact of blood glucose levels on diabetes diagnosis.</p>
<p>First let’s make a scatterplot. Could there be a relationship?</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb256-1"><a href="module-3-generalized-linear-models.html#cb256-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(<span class="at">x =</span> glucose, <span class="at">y =</span> <span class="fu">factor</span>(diabetes)))</span>
<span id="cb256-2"><a href="module-3-generalized-linear-models.html#cb256-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 5 rows containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<p><img src="_main_files/figure-html/plot-diabetes-vs-glucose-1.png" width="672" /></p>
<p>Could this be fit with a linear model?</p>
<p>Is there a continuous line that could reasonably fit the data?</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb258-1"><a href="module-3-generalized-linear-models.html#cb258-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(<span class="at">x =</span> glucose, <span class="at">y =</span> <span class="fu">factor</span>(diabetes)))</span>
<span id="cb258-2"><a href="module-3-generalized-linear-models.html#cb258-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning: Removed 5 rows containing non-finite outside the scale range
## (`stat_smooth()`).</code></pre>
<pre><code>## Warning: Removed 5 rows containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<p><img src="_main_files/figure-html/unnamed-chunk-72-1.png" width="672" /></p>
<p>This is a situation where a logistic regression model would be an appropriate choice because of the binary outcome.</p>
<p>We’re going to use <code>glm()</code> to fit a model to these data:</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb262-1"><a href="module-3-generalized-linear-models.html#cb262-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span>glucose, </span>
<span id="cb262-2"><a href="module-3-generalized-linear-models.html#cb262-2" tabindex="-1"></a>    dat,</span>
<span id="cb262-3"><a href="module-3-generalized-linear-models.html#cb262-3" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span> <span class="co"># set to model binary outcome</span></span>
<span id="cb262-4"><a href="module-3-generalized-linear-models.html#cb262-4" tabindex="-1"></a>)</span>
<span id="cb262-5"><a href="module-3-generalized-linear-models.html#cb262-5" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose, family = &quot;binomial&quot;, 
##     data = dat)
## 
## Coefficients:
##              Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept) -5.715088   0.438100  -13.04   &lt;2e-16 ***
## glucose      0.040634   0.003382   12.01   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 986.70  on 762  degrees of freedom
## Residual deviance: 786.56  on 761  degrees of freedom
##   (5 observations deleted due to missingness)
## AIC: 790.56
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>Which factors explain a diabetes diagnosis?</p>
<p>What if we include a couple other factors?</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb264-1"><a href="module-3-generalized-linear-models.html#cb264-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span> glucose <span class="sc">+</span> pregnant <span class="sc">+</span> age <span class="sc">+</span> pressure <span class="sc">+</span> triceps,</span>
<span id="cb264-2"><a href="module-3-generalized-linear-models.html#cb264-2" tabindex="-1"></a>    dat,</span>
<span id="cb264-3"><a href="module-3-generalized-linear-models.html#cb264-3" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb264-4"><a href="module-3-generalized-linear-models.html#cb264-4" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose + pregnant + age + pressure + 
##     triceps, family = &quot;binomial&quot;, data = dat)
## 
## Coefficients:
##               Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept) -7.6685706  0.8451262  -9.074  &lt; 2e-16 ***
## glucose      0.0365926  0.0041105   8.902  &lt; 2e-16 ***
## pregnant     0.0972493  0.0417560   2.329 0.019860 *  
## age          0.0249707  0.0135945   1.837 0.066235 .  
## pressure    -0.0009321  0.0099322  -0.094 0.925230    
## triceps      0.0420045  0.0117044   3.589 0.000332 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 678.40  on 533  degrees of freedom
## Residual deviance: 494.38  on 528  degrees of freedom
##   (234 observations deleted due to missingness)
## AIC: 506.38
## 
## Number of Fisher Scoring iterations: 5</code></pre>
<div class="rmd-caution">
<p>
Note: This morning’s session only intends to introduce you to fitting
non-linear models.
</p>
<p>
In practice you may need to do more work to test multiple models to
ascertain best fits your data, using measures such as goodness-of-fit.
You will also likely compute odds ratio (odds of increased Y per unit
increase X), which is out of scope for the current tutorial.
</p>
<p>
We strongly recommend that you learn these topics before applying
these methods to your own data.
</p>
</div>
</div>
<div id="bonus-exercise-2" class="section level3 unnumbered hasAnchor">
<h3>Bonus Exercise<a href="module-3-generalized-linear-models.html#bonus-exercise-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now let’s apply the ideas above to a dataset for classifying a breast cell as being either benign or malignant.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb266-1"><a href="module-3-generalized-linear-models.html#cb266-1" tabindex="-1"></a><span class="fu">data</span>(BreastCancer)</span>
<span id="cb266-2"><a href="module-3-generalized-linear-models.html#cb266-2" tabindex="-1"></a>bc <span class="ot">&lt;-</span> BreastCancer</span>
<span id="cb266-3"><a href="module-3-generalized-linear-models.html#cb266-3" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="co"># altered for current lab</span></span>
<span id="cb266-4"><a href="module-3-generalized-linear-models.html#cb266-4" tabindex="-1"></a>    bc[,k] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(bc[,k]) </span>
<span id="cb266-5"><a href="module-3-generalized-linear-models.html#cb266-5" tabindex="-1"></a><span class="fu">head</span>(bc)</span></code></pre></div>
<pre><code>##        Id Cl.thickness Cell.size Cell.shape Marg.adhesion Epith.c.size
## 1 1000025            5         1          1             1            2
## 2 1002945            5         4          4             5            7
## 3 1015425            3         1          1             1            2
## 4 1016277            6         8          8             1            3
## 5 1017023            4         1          1             3            2
## 6 1017122            8        10         10             8            7
##   Bare.nuclei Bl.cromatin Normal.nucleoli Mitoses     Class
## 1           1           3               1       1    benign
## 2          10           3               2       1    benign
## 3           2           3               1       1    benign
## 4           4           3               7       1    benign
## 5           1           3               1       1    benign
## 6          10           9               7       1 malignant</code></pre>
<p>Learn more about the dataset:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb268-1"><a href="module-3-generalized-linear-models.html#cb268-1" tabindex="-1"></a>?BreastCancer</span></code></pre></div>
<p>For your exercise, answer the following questions:</p>
<ul>
<li>Is there missing data?</li>
<li>Which columns are</li>
<li>Use plots to explore the relationship between explanatory variables.</li>
<li>Fit a logistic model to identify which factors explain class (benign vs. malignant).</li>
</ul>
</div>
<div id="module-3-bonus-exercise-results" class="section level3 unnumbered hasAnchor">
<h3>Module 3: Bonus Exercise Results<a href="module-3-generalized-linear-models.html#module-3-bonus-exercise-results" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb269"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb269-1"><a href="module-3-generalized-linear-models.html#cb269-1" tabindex="-1"></a><span class="fu">data</span>(BreastCancer)</span>
<span id="cb269-2"><a href="module-3-generalized-linear-models.html#cb269-2" tabindex="-1"></a>bc <span class="ot">&lt;-</span> BreastCancer</span>
<span id="cb269-3"><a href="module-3-generalized-linear-models.html#cb269-3" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="co"># altered for current lab</span></span>
<span id="cb269-4"><a href="module-3-generalized-linear-models.html#cb269-4" tabindex="-1"></a>    bc[,k] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(bc[,k]) </span>
<span id="cb269-5"><a href="module-3-generalized-linear-models.html#cb269-5" tabindex="-1"></a><span class="fu">head</span>(bc)</span></code></pre></div>
<pre><code>##        Id Cl.thickness Cell.size Cell.shape Marg.adhesion Epith.c.size
## 1 1000025            5         1          1             1            2
## 2 1002945            5         4          4             5            7
## 3 1015425            3         1          1             1            2
## 4 1016277            6         8          8             1            3
## 5 1017023            4         1          1             3            2
## 6 1017122            8        10         10             8            7
##   Bare.nuclei Bl.cromatin Normal.nucleoli Mitoses     Class
## 1           1           3               1       1    benign
## 2          10           3               2       1    benign
## 3           2           3               1       1    benign
## 4           4           3               7       1    benign
## 5           1           3               1       1    benign
## 6          10           9               7       1 malignant</code></pre>
<p>Explore missingness:</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb271-1"><a href="module-3-generalized-linear-models.html#cb271-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(plotrix))</span>
<span id="cb271-2"><a href="module-3-generalized-linear-models.html#cb271-2" tabindex="-1"></a></span>
<span id="cb271-3"><a href="module-3-generalized-linear-models.html#cb271-3" tabindex="-1"></a><span class="co">#&#39; show data missingness as a chequered matrix</span></span>
<span id="cb271-4"><a href="module-3-generalized-linear-models.html#cb271-4" tabindex="-1"></a><span class="co">#&#39; </span></span>
<span id="cb271-5"><a href="module-3-generalized-linear-models.html#cb271-5" tabindex="-1"></a><span class="co">#&#39; @param x (matrix) data matrix.</span></span>
<span id="cb271-6"><a href="module-3-generalized-linear-models.html#cb271-6" tabindex="-1"></a><span class="co">#&#39; @param outFile (char) path to file for printing graph</span></span>
<span id="cb271-7"><a href="module-3-generalized-linear-models.html#cb271-7" tabindex="-1"></a><span class="co">#&#39; @param wd (numeric) width in inches</span></span>
<span id="cb271-8"><a href="module-3-generalized-linear-models.html#cb271-8" tabindex="-1"></a><span class="co">#&#39; @param ht (numeric) height in inches</span></span>
<span id="cb271-9"><a href="module-3-generalized-linear-models.html#cb271-9" tabindex="-1"></a><span class="co">#&#39; @return plots missingness matrix to file</span></span>
<span id="cb271-10"><a href="module-3-generalized-linear-models.html#cb271-10" tabindex="-1"></a><span class="co">#&#39; @import plotrix</span></span>
<span id="cb271-11"><a href="module-3-generalized-linear-models.html#cb271-11" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb271-12"><a href="module-3-generalized-linear-models.html#cb271-12" tabindex="-1"></a>plotMissMat <span class="ot">&lt;-</span> <span class="cf">function</span>(x,<span class="at">xlab=</span><span class="st">&quot;columns&quot;</span>,</span>
<span id="cb271-13"><a href="module-3-generalized-linear-models.html#cb271-13" tabindex="-1"></a>        <span class="at">ylab=</span><span class="st">&quot;rows&quot;</span>,<span class="at">border=</span><span class="cn">NA</span>) {</span>
<span id="cb271-14"><a href="module-3-generalized-linear-models.html#cb271-14" tabindex="-1"></a>    </span>
<span id="cb271-15"><a href="module-3-generalized-linear-models.html#cb271-15" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(x)</span>
<span id="cb271-16"><a href="module-3-generalized-linear-models.html#cb271-16" tabindex="-1"></a>    <span class="fu">class</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;numeric&quot;</span></span>
<span id="cb271-17"><a href="module-3-generalized-linear-models.html#cb271-17" tabindex="-1"></a>    <span class="fu">color2D.matplot</span>(x,<span class="at">show.values=</span><span class="cn">FALSE</span>,<span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb271-18"><a href="module-3-generalized-linear-models.html#cb271-18" tabindex="-1"></a>        <span class="at">cs1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs2=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs3=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">border=</span>border,</span>
<span id="cb271-19"><a href="module-3-generalized-linear-models.html#cb271-19" tabindex="-1"></a>        <span class="at">cex=</span><span class="fl">0.8</span>,</span>
<span id="cb271-20"><a href="module-3-generalized-linear-models.html#cb271-20" tabindex="-1"></a>        <span class="at">xlab=</span>xlab,<span class="at">ylab=</span>ylab)</span>
<span id="cb271-21"><a href="module-3-generalized-linear-models.html#cb271-21" tabindex="-1"></a>}</span></code></pre></div>
<p>Explore missingness:</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb272-1"><a href="module-3-generalized-linear-models.html#cb272-1" tabindex="-1"></a><span class="fu">plotMissMat</span>(bc)</span></code></pre></div>
<p><img src="_main_files/figure-html/ex3-miss2-1.png" width="672" /></p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb273-1"><a href="module-3-generalized-linear-models.html#cb273-1" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(bc))</span></code></pre></div>
<pre><code>##              Id    Cl.thickness       Cell.size      Cell.shape   Marg.adhesion 
##               0               0               0               0               0 
##    Epith.c.size     Bare.nuclei     Bl.cromatin Normal.nucleoli         Mitoses 
##               0              16               0               0               0 
##           Class 
##               0</code></pre>
<p>Plot relationship between variables</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb275-1"><a href="module-3-generalized-linear-models.html#cb275-1" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb275-2"><a href="module-3-generalized-linear-models.html#cb275-2" tabindex="-1"></a></span>
<span id="cb275-3"><a href="module-3-generalized-linear-models.html#cb275-3" tabindex="-1"></a><span class="fu">ggplot</span>(bc,<span class="fu">aes</span>(Cell.size,Normal.nucleoli)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/ex3-plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb276-1"><a href="module-3-generalized-linear-models.html#cb276-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Normal.nucleoli))</span></code></pre></div>
<p><img src="_main_files/figure-html/ex3-plot-2.png" width="672" /></p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb277-1"><a href="module-3-generalized-linear-models.html#cb277-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Cl.thickness))</span></code></pre></div>
<p><img src="_main_files/figure-html/ex3-plot-3.png" width="672" /></p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb278-1"><a href="module-3-generalized-linear-models.html#cb278-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Bare.nuclei))</span></code></pre></div>
<pre><code>## Warning: Removed 16 rows containing non-finite outside the scale range
## (`stat_boxplot()`).</code></pre>
<p><img src="_main_files/figure-html/ex3-plot-4.png" width="672" /></p>
<p>Fit a binary outcome model:</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb280-1"><a href="module-3-generalized-linear-models.html#cb280-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(</span>
<span id="cb280-2"><a href="module-3-generalized-linear-models.html#cb280-2" tabindex="-1"></a>    Class <span class="sc">~</span> Cl.thickness <span class="sc">+</span> Bare.nuclei <span class="sc">+</span> Normal.nucleoli <span class="sc">+</span> Mitoses <span class="sc">+</span> Bl.cromatin, </span>
<span id="cb280-3"><a href="module-3-generalized-linear-models.html#cb280-3" tabindex="-1"></a>    bc,</span>
<span id="cb280-4"><a href="module-3-generalized-linear-models.html#cb280-4" tabindex="-1"></a>    <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb280-5"><a href="module-3-generalized-linear-models.html#cb280-5" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = Class ~ Cl.thickness + Bare.nuclei + Normal.nucleoli + 
##     Mitoses + Bl.cromatin, family = &quot;binomial&quot;, data = bc)
## 
## Coefficients:
##                 Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)     -9.69635    0.98763  -9.818  &lt; 2e-16 ***
## Cl.thickness     0.63413    0.12267   5.169 2.35e-07 ***
## Bare.nuclei      0.50988    0.08228   6.197 5.76e-10 ***
## Normal.nucleoli  0.35765    0.10040   3.562 0.000368 ***
## Mitoses          0.52677    0.27923   1.887 0.059223 .  
## Bl.cromatin      0.61398    0.14225   4.316 1.59e-05 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 884.35  on 682  degrees of freedom
## Residual deviance: 120.81  on 677  degrees of freedom
##   (16 observations deleted due to missingness)
## AIC: 132.81
## 
## Number of Fisher Scoring iterations: 8</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2-dimensionality-reduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/cbw-dev/AUR_2024/blob/main/030-module-3.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
